### PR TITLE
feature/2007_stepper

### LINKED ISSUE #2007: Progressbar für Wizards
Es wäre toll wenn die Patterns Library eine Progressbar für Wizards anbieten würde. Aktuell verwenden wir die Komponente vom ESales Team: https://github.com/axa-ch/esales-ui-components/

Die ESales Komponente könnte hier als Inspiration dienen.

Die Progressbar sollte folgende Features beinhalten:
* Beliebige Anzahl von Schritten 
* Markierung bereits erledigter Schritte
* Fortschrittsanzeige innerhalb eines Schrittes (Wenn ein Schritt z.B. Formular heisst aber aus mehreren Unterschrit
ten besteht.
* Stepper State sollte von aussen gemanaged werden können

Bild:
Rot: Erledigter Step
Blau: Anzeige das der Step erledigt wurde (inkl. "rotem" balken unten)
Grün: Wenn es Unterschritte innerhalb eines Schrittes gibt (z.B. besteht das Formular aus mehreren Seiten (Schritten) und 1/4 Seiten wurde bereits ausgefüllt)

Progressbar

### PR DESCRIPTION
**Pull request by raphael-suter**
Fixes #2007 

# Merge Checklist:
- [x] Code selfreview has been performed ([Best Practices](https://github.com/axa-ch/patterns-library/blob/develop/CONTRIBUTION.md#best-practices)).
- [x] Tests are sufficient.
- [x] Modifications still work in IE.
- [x] Documentation is up to date.
- [x] Changelog is up to date.
- [x] Typescript typings for properties are up to date.
- [x] Designers approved changes or no approval needed.

### CODE REVIEW

File: src/components/20-molecules/stepper/README.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+## Properties
+
+### steps
+
+The Array attribute `steps` specifies the descriptions of the steps. The length of this array equals the number of steps.
+
+### stepActive
+
+The Number attribute `stepActive` defines the currently active step.
```
Comment (markus-walther) [inline src/components/20-molecules/stepper/README.md]: Number -> integer-valued

File: src/components/20-molecules/stepper/README.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    <script src="node_modules/@axa-ch/stepper/dist/index.js"></script>
+  </body>
+</html>
+```
+
+## Properties
+
+### steps
+
+The Array attribute `steps` specifies the descriptions of the steps. The length of this array equals the number of steps.
```
Comment (markus-walther) [inline src/components/20-molecules/stepper/README.md]: Better: The **array-valued** attribute ... of the steps, **each step** being described by a non-empty label string.

File: src/components/20-molecules/stepper/README.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+### stepActive
+
+The Number attribute `stepActive` defines the currently active step.
+
+### stepProgress
+
+The Number attribute `stepProgress` defines the progress of the currently active step.
```
Comment (markus-walther) [inline src/components/20-molecules/stepper/README.md]: It's unclear what that means, since you nowhere define 'progress'. Make this crystal-clear...

Comment (markus-walther) [reply src/components/20-molecules/stepper/README.md]: And Number -> numeric or floating-point

File: src/components/20-molecules/stepper/index.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  constructor() {
+    super();
+    applyDefaults(this);
+  }
+
+  render() {
+    const { min, max, floor } = Math;
+    let { steps, stepActive, stepProgress } = this;
+
+    stepActive = floor(max(stepActive, 1));
```
Comment (markus-walther) [inline src/components/20-molecules/stepper/index.js:39]: comment: ensure it's a positive integer

File: src/components/20-molecules/stepper/index.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    stepProgress = max(min(stepProgress, 1), 0);
```
Comment (markus-walther) [inline src/components/20-molecules/stepper/index.js:42]: comment: bound stepProgress between 0 and 1, inclusive

File: src/components/20-molecules/stepper/index.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+    steps = steps.map((text, index) => {
+      const currentStep = index + 1;
+
+      let state = 'active';
+      let symbol = html`
+        <span class="m-stepper__circle js-stepper__circle">${currentStep}</span>
+      `;
+
+      if (currentStep < stepActive) {
```
Comment (markus-walther) [inline src/components/20-molecules/stepper/index.js:54]: add comments, esp. to this logic

File: src/components/20-molecules/stepper/index.js [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+      }
+
+      return html`
+        <div class="m-stepper__step m-stepper__step--${state} js-stepper__step">
+          ${symbol}<span class="m-stepper__text js-stepper__text">${text}</span>
+        </div>
+      `;
+    });
+
+    const progress = (100 / steps.length) * (stepActive - 1 + stepProgress);
```
Comment (markus-walther) [inline src/components/20-molecules/stepper/index.js:71]: comment: convert stepper state to progress-bar length (in percent)

File: src/components/20-molecules/stepper/index.scss [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+  }
+
+  &__progress {
+    height: 2px;
+    max-width: 100%;
+    background: $color-axa-blue;
+  }
+}
+
+@include breakpoint($mediaquery-sm-up) {
```
Comment (markus-walther) [inline src/components/20-molecules/stepper/index.scss:91]: good!

File: src/components/20-molecules/stepper/README.md (hunk 3 above)
Comment (markus-walther) [reply src/components/20-molecules/stepper/README.md]: "Legal values are between 0 and 1, inclusive (e.g. 0.5)."

File: src/components/20-molecules/stepper/story.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    'Fertig',
+  ]);
+
+  const stepActive = number('stepActive', 2);
+  const stepProgress = number('stepProgress', 0.5);
+
+  const template = html`
+    <axa-stepper
+      .steps=${steps}
+      stepactive=${stepActive || 0}
```
Comment (markus-walther) [inline src/components/20-molecules/stepper/story.js]: You use lit-html for the template. Why then go through attributes? `.stepActive/Progress={...}`!

File: src/components/20-molecules/stepper/ui.test.js [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+
+  const $axaElemShadowEl = $axaElemShadow.find(CLASS);
+  await t.expect($axaElemShadowEl.exists).ok();
+});
+
+test('should render correct amount of steps', async t => {
+  const amountOfSteps = ClientFunction(
+    () => {
+      const shadow = document.querySelector(TAG).shadowRoot;
+      return shadow.querySelectorAll('.js-stepper__step').length;
```
Comment (markus-walther) [inline src/components/20-molecules/stepper/ui.test.js:29]: `js-stepper__step` - good!

File: src/components/20-molecules/stepper/ui.test.js [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+    },
+    { dependencies: { TAG } }
+  );
+
+  await t.expect(await amountOfSteps()).eql(4);
+});
+
+test('should display correct step number', async t => {
+  const getStepNumber = ClientFunction(
+    index => {
```
Comment (markus-walther) [inline src/components/20-molecules/stepper/ui.test.js:39]: clever use of _parametrized_ ClientFunction!

File: src/components/20-molecules/stepper/ui.test.js [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+      const shadow = document.querySelector(TAG).shadowRoot;
+      const step = shadow.querySelectorAll('.js-stepper__circle')[index];
+
+      return step.textContent;
+    },
+    { dependencies: { TAG } }
+  );
+
+  await t.expect(await getStepNumber(0)).eql('2');
+  await t.expect(await getStepNumber(1)).eql('3');
```
Comment (markus-walther) [inline src/components/20-molecules/stepper/ui.test.js:49]: DRY violation - could you use an array of expected outcomes (const EXPECTED = [2,3,4]) and a loop?

File: src/components/20-molecules/stepper/ui.test.js [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+    index => {
+      const shadow = document.querySelector(TAG).shadowRoot;
+      const step = shadow.querySelectorAll('.js-stepper__text')[index];
+
+      return step.textContent;
+    },
+    { dependencies: { TAG } }
+  );
+
+  await t.expect(await getStepTitle(0)).eql('Angaben');
```
Comment (markus-walther) [inline src/components/20-molecules/stepper/ui.test.js:64]: similar DRY violation, perhaps similar solution to above suggestion?

File: src/components/20-molecules/stepper/ui.test.js [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+      } else if (circleBackground === 'rgb(255, 255, 255)') {
+        return 2;
+      }
+
+      return 3;
+    },
+    { dependencies: { TAG } }
+  );
+
+  await t.expect(await getStepState(0)).eql(0);
```
Comment (markus-walther) [inline src/components/20-molecules/stepper/ui.test.js:96]: DRY violation, see above

File: src/components/20-molecules/stepper/ui.test.js [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+  );
+
+  await setStepActive(0);
+  await t.expect(await getStepState()).ok();
+
+  await setStepActive(1);
+  await t.expect(await getStepState()).ok();
+});
+
+test('should round stepProgress up to 0 if it is smaller', async t => {
```
Comment (markus-walther) [inline src/components/20-molecules/stepper/ui.test.js]: wording: this is not rounding. Limit? Bound by?

File: src/components/20-molecules/stepper/ui.test.js [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+      return progress.style.width;
+    },
+    { dependencies: { TAG } }
+  );
+
+  await setStepProgress(-1);
+  await t.expect(await getProgress()).eql('25%');
+});
+
+test('should round stepProgress down to 1 if it is greater', async t => {
```
Comment (markus-walther) [inline src/components/20-molecules/stepper/ui.test.js]: ditto

**Review by markus-walther — CHANGES_REQUESTED** [review]
Good work, Raphi! Some small findings, shouldn't be to difficult to address.

Cheers, Markus

File: src/components/20-molecules/stepper/README.md (hunk 3 above)
Comment (raphael-suter) [reply src/components/20-molecules/stepper/README.md]: I think it should be clear. What would be your suggestion?

**Review by 01231 — CHANGES_REQUESTED** [review]
Numbers don't seem to be aligned perfectly. (More space on the top en right side)
Screenshot 2020-10-28 at 15 20 50

File: src/components/20-molecules/stepper/README.md [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+
+The array-valued attribute `steps` specifies the descriptions of the steps, each step being described by a non-empty label string.
+
+### stepActive
+
+The integer-valued attribute `stepActive` defines the currently active step.
+
+### stepProgress
+
+The numeric attribute `stepProgress` defines the progress of the currently active step. Legal values are between 0 and 1, inclusive (e.g. 0.5).
```
Comment (01231) [inline src/components/20-molecules/stepper/README.md]: Would rephrase like this: Legal values are between 0 and 1, including decimals (e.g. 0.5).

**Review by 01231 — DISMISSED** [review]
Looks very good. 

Only thing that bothers me know is, that the knobs of `stepProgress` has this little arrows which are not very useful, because only values between 0 and 1 are valid. If you click once on it, it directly goes to 1.5 or -0.5.

File: src/components/20-molecules/stepper/README.md (hunk 17 above)
Comment (markus-walther) [reply src/components/20-molecules/stepper/README.md]: Even better, since this is adressed to JS developers: Legal values are floating-point numbers between 0 and 1, inclusive (e.g. 0.5).

**Review by markus-walther — CHANGES_REQUESTED** [review]
Two small nitpicks only:

- floating-point numbers, i.e. correct technical terminology
- sharpen the term progress to refer to the UI's progress bar somehow

**Review by 01231 — APPROVED** [review]

File: src/components/20-molecules/stepper/README.md [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+
+The array-valued attribute `steps` specifies the descriptions of the steps, each step being described by a non-empty label string.
+
+### stepActive
+
+The integer-valued attribute `stepActive` defines the currently active step.
+
+### stepProgress
+
+The numeric attribute `stepProgress` defines the fraction, to which the UI progress bar extends into the currently active step. Legal values are floating-point numbers between 0 and 1, inclusive (e.g. 0.5).
```
Comment (markus-walther) [inline src/components/20-molecules/stepper/README.md:68]: great, sounds perfect now!

**Review by markus-walther — APPROVED** [review]
LGTM
